### PR TITLE
Expose `matches`, `ranges` and `split`

### PR DESCRIPTION
**Pull request by itingliu**
Publicize these API per the String Processing Algorithms proposal. The proposed
ones return generic `Collection`, empowered by SE-0346. For now we'll wrap the
results with a concrete `Array` until the language feature is ready.

### CODE REVIEW

**Comment by itingliu** [comment]
@swift-ci please test

File: Sources/_StringProcessing/Algorithms/Algorithms/Ranges.swift [hunk 1]
```diff
@@ -231,4 +243,17 @@ extension BidirectionalCollection where SubSequence == Substring {
...
+  /// Finds and returns the ranges of the all occurrences of a given sequence
+  /// within the collection.
+  /// - Parameter regex: The regex to search for.
+  /// - Returns: A collection or ranges in the receiver of all occurrences of
+  /// `regex`. Returns an empty collection if `regex` is not found.
+  @available(SwiftStdlib 5.7, *)
+  public func ranges<R: RegexComponent>(
+    of regex: R
+  ) -> [Range<Index>] {
+    ranges(of: RegexConsumer(regex)).map{ $0 }
```
Comment (milseman) [inline Sources/_StringProcessing/Algorithms/Algorithms/Ranges.swift:257]:
```suggestion
    Array(ranges(of: RegexConsumer(regex)))
```

File: Sources/_StringProcessing/Algorithms/Algorithms/Split.swift [hunk 2]
```diff
@@ -233,16 +233,24 @@ extension BidirectionalCollection where Element: Equatable {
...
   /// - Returns: A collection of subsequences, split from this collection's
   /// elements.
-  func split<S: Sequence>(
+  @available(SwiftStdlib 5.7, *)
+  public func split<S: Sequence>(
     by separator: S
-  ) -> SplitCollection<ZSearcher<Self>> where S.Element == Element {
-    split(by: ZSearcher(pattern: Array(separator), by: ==))
+  ) -> [SubSequence] where S.Element == Element {
+    split(by: ZSearcher(pattern: Array(separator), by: ==)).map{ $0 }
```
Comment (milseman) [inline Sources/_StringProcessing/Algorithms/Algorithms/Split.swift:253]:
```suggestion
    Array(split(by: ZSearcher(pattern: Array(separator), by: ==)))
```

File: Sources/_StringProcessing/Algorithms/Algorithms/Split.swift [hunk 3]
```diff
@@ -299,4 +302,16 @@ extension BidirectionalCollection where SubSequence == Substring {
...
+  // FIXME: Return `some Collection<Substring>` for SE-0346 
+  /// Returns the longest possible subsequences of the collection, in order,
+  /// around elements equal to the given separator.
+  /// - Parameter separator: A regex describing elements to be split upon.
+  /// - Returns: A collection of substrings, split from this collection's
+  /// elements.
+  public func split<R: RegexComponent>(
+    by separator: R
+  ) -> [SubSequence] {
+    split(by: RegexConsumer(separator)).map{ $0 }
```
Comment (milseman) [inline Sources/_StringProcessing/Algorithms/Algorithms/Split.swift:315]:
```suggestion
    Array(split(by: RegexConsumer(separator)))
```

**Review by milseman — APPROVED** [review]
LGTM

**Comment by itingliu** [comment]
@swift-ci please test

**Pull request merged**